plugin/backend_linux.go

```diff
@@ -797,6 +813,262 @@ func (pm *Manager) CreateFromContext(ctx context.Context, tarCtx io.ReadCloser,
+}
+
+func safePath(base, path string) (string, error) {
+	return symlink.FollowSymlinkInScope(filepath.Join(base, path), base)
+}
+
+// LoadPlugin loads a plugin from input tar archive
+func (pm *Manager) LoadPlugin(input io.ReadCloser, outStream io.Writer, quiet bool) (err error) {
+	outStream = streamformatter.NewStdoutWriter(outStream)
+
+	// 1. create the tempdir used to untar and then load
+	tmpDir, err := ioutil.TempDir("", "docker-plugin-import-")
+	if err != nil {
+		return err
+	}
+	defer os.RemoveAll(tmpDir)
+
+	// 2. untar the given tar
+	if err := chrootarchive.Untar(input, tmpDir, nil); err != nil {
+		return err
+	}
+
+	// 3. open manifest.
+	manifestPath, err := safePath(tmpDir, manifestFileName)
+	if err != nil {
+		return err
+	}
+	manifestFile, err := os.Open(manifestPath)
+	if err != nil {
+		return err
+	}
+	defer manifestFile.Close()
+
+	// 4. decode the json file into a struct
+	var manifest manifestItem
+	if err := json.NewDecoder(manifestFile).Decode(&manifest); err != nil {
+		return err
+	}
+
+	// 5. read config file
+	configPath, err := safePath(tmpDir, manifest.Config)
+	if err != nil {
+		return err
+	}
+	rdr, err := os.Open(configPath)
+	if err != nil {
+		return err
+	}
+	defer rdr.Close()
+
+	// write config to blobStore.
+	configBlob, err := pm.blobStore.New()
+	if err != nil {
+		return err
+	}
+	defer configBlob.Close()
+
+	configRdr := io.TeeReader(rdr, configBlob)
+	configBytes, err := ioutil.ReadAll(configRdr)
+	if err != nil {
+		return err
+	}
+	_, err = configBlob.Commit()
+	if err != nil {
+		return err
+	}
+
+	var config types.PluginConfig
+	if err := json.Unmarshal(configBytes, &config); err != nil {
+		return err
+	}
+	privileges, err := computePrivileges(config)
+	if err != nil {
+		return err
+	}
+
+	// 6. read rootfs
+	compressedTar, err := safePath(tmpDir, manifest.Layers[0])
+	if err != nil {
+		return err
+	}
+	// open the compressedTar
+	compressedTarRdr, err := os.Open(compressedTar)
+	if err != nil {
+		return err
+	}
+	defer compressedTarRdr.Close()
+
+	// XXX: Instead of using the blobStore interface to create, write and commit
+	// to blobstore, why not simply copy the files and rename them?
+
+	// write rootfs to blobStore
+	rootfsBlob, err := pm.blobStore.New()
+	if err != nil {
+		return err
+	}
+	rootfsRdr := io.TeeReader(compressedTarRdr, rootfsBlob)
+	_, err = ioutil.ReadAll(rootfsRdr)
+	if err != nil {
+		return err
+	}
+	_, err = rootfsBlob.Commit()
+	if err != nil {
+		return err
+	}
+
+	// file seek to the beginning for DecompressStream's reader
+	compressedTarRdr.Seek(0, 0)
+	// decompress the tar.gz archive.
+	decompressedTarRdr, err := archive.DecompressStream(compressedTarRdr)
+	if err != nil {
+		return err
+	}
+	// create a tmpdir to extract the decompressedTarRdr
+	tmpRootFS, err := ioutil.TempDir(tmpDir, ".rootfs")
+	if err != nil {
+		return err
+	}
+	defer os.RemoveAll(tmpRootFS)
+	if err = chrootarchive.Untar(decompressedTarRdr, tmpRootFS, nil); err != nil {
```

Comment:
`ApplyLayer` needs to be used here. Could probably also combine into utility function to share this code with install.